step-generation/src/types.ts

```diff
@@ -637,7 +637,10 @@ export interface TimelineFrame {
       }
     }
     pipettes: {
-      [pipetteId: string]: boolean // true if pipette has tip(s)
+      [pipetteId: string]: {
+        hasTip: boolean
+        attachedTipURI: string | null
```

Comment:
slight nit, this is actually the tiprack uri that the tip is a part of right? i think the naming is a bit confusing since it seems like the tip itself has a URI

```suggestion
        tiprackURI: string | null
```